core/src/main/java/com/graphhopper/matrix/DistanceMatrixService.java

```diff
@@ -0,0 +1,186 @@
+package com.graphhopper.matrix;
+
+import com.graphhopper.GHResponse;
+import com.graphhopper.GraphHopper;
+import com.graphhopper.matrix.algorithm.MatrixAlgorithm;
+import com.graphhopper.matrix.algorithm.MatrixAlgorithmFactory;
+import com.graphhopper.matrix.algorithm.SimpleMatrixAlgorithmFactory;
+import com.graphhopper.routing.AlgorithmOptions;
+import com.graphhopper.routing.QueryGraph;
+import com.graphhopper.routing.RoutingAlgorithmFactory;
+import com.graphhopper.routing.ch.CHAlgoFactoryDecorator;
+import com.graphhopper.routing.ch.PrepareContractionHierarchies;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.routing.util.TraversalMode;
+import com.graphhopper.routing.util.Weighting;
+import com.graphhopper.storage.CHGraph;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.storage.index.QueryResult;
+import com.graphhopper.util.shapes.GHPoint;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Provides a Distance Matrix API
+ *
+ * @author Pascal Büttiker
+ */
+public class DistanceMatrixService {
+
+    private final GraphHopper hopper; // TODO after GraphHopper refactoring, this dependency should go away
+    private final MatrixAlgorithmFactory matrixAlgorithmFactory;
+
+
+    public DistanceMatrixService(GraphHopper hopper){
+        this(hopper, new SimpleMatrixAlgorithmFactory());
+    }
+
+    public DistanceMatrixService(GraphHopper hopper, MatrixAlgorithmFactory matrixAlgorithmFactory){
+        this.hopper = hopper;
+        this.matrixAlgorithmFactory = matrixAlgorithmFactory;
+    }
+
+    /**
+     * Calculates a distance matrix based on the given request.
+     */
+    public GHMatrixResponse calculateMatrix(GHMatrixRequest request){
+
+
+        AlgorithmOptions.Builder algoOptsBuilder =  buildOptions(request, hopper);
+
+
+        AlgorithmOptions tmpOptions = algoOptsBuilder.build();
+
+
+        Graph routingGraph = hopper.getGraphHopperStorage();
+        if(hopper.isCHEnabled()) {
+            routingGraph = hopper.getGraphHopperStorage().getGraph(CHGraph.class, tmpOptions.getWeighting());
+        }
+
+        QueryGraph queryGraph = new QueryGraph(routingGraph);
+        Weighting weighting = hopper.createTurnWeighting(tmpOptions.getWeighting(), queryGraph, tmpOptions.getFlagEncoder());
+        AlgorithmOptions algoOpts = algoOptsBuilder.weighting(weighting).build();
+
+        List<QueryResult> originQNodes = lookupNodes(request.getOrigins(), algoOpts.getFlagEncoder());
+        List<QueryResult> destinationQNodes = lookupNodes(request.getDestinations(), algoOpts.getFlagEncoder());
+
+        List<QueryResult> merged = new ArrayList<>(originQNodes);
+        merged.addAll(destinationQNodes);
+
+        queryGraph.lookup(merged);
+
+        List<Integer> originNodes =  mapToNodes(originQNodes);
+        List<Integer> destinationNodes =  mapToNodes(destinationQNodes);
```

Comment:
Well I need the QueryResults due to the necessary `queryGraph.lookup(merged);` to properly initialise the queryGraph.

But since this code is almost 1:1 identical to the GraphHopper default route Graph initialisation, I really think this should be refactored in a shared Graph-Preparation class.